substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -549,39 +551,36 @@ where
 		});
 
 		if let Some(current_best_block) = self.current_best_block {
-			// The best reported block is in the pruned list. Report a new best block.
+			// There are two cases when the best block is on fork:
+			//  - The best block is in the pruned list and is reported immediately
+			//  - The best block is on another fork that will be pruned in the future.
+			//
+			// For both cases, we need to report a new best block.
+
+			// Case 1. The best reported block is in the pruned list.
 			let is_in_pruned_list =
 				pruned_block_hashes.iter().any(|hash| *hash == current_best_block);
-			// The block is not the last finalized block.
-			//
-			// It can be either:
-			//  - a descendant of the last finalized block
-			//  - a block on a fork that will be pruned in the future.
-			//
-			// In those cases, we emit a new best block.
-			let is_not_last_finalized = current_best_block != last_finalized;
-
-			if is_in_pruned_list || is_not_last_finalized {
-				// We need to generate a best block event.
-				let best_block_hash = self.client.info().best_hash;
-
-				// Defensive check against state missmatch.
-				if best_block_hash == current_best_block {
-					// The client doest not have any new information about the best block.
-					// The information from `.info()` is updated from the DB as the last
-					// step of the finalization and it should be up to date.
-					// If the info is outdated, there is nothing the RPC can do for now.
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Client does not contain different best block",
-						self.sub_id,
-					);
-				} else {
-					// The RPC needs to also submit a new best block changed before the
-					// finalized event.
-					self.current_best_block = Some(best_block_hash);
-					events
-						.push(FollowEvent::BestBlockChanged(BestBlockChanged { best_block_hash }));
+			if is_in_pruned_list {
+				self.current_best_block = Some(last_finalized);
+				events.push(FollowEvent::BestBlockChanged(BestBlockChanged {
+					best_block_hash: last_finalized,
+				}));
+			} else {
+				// Case 2. Check if the best block is a descendant of the last finalized block.
+				let ancestor = sp_blockchain::lowest_common_ancestor(
+					&*self.client,
+					last_finalized,
+					current_best_block,
+				)?;
+
+				let is_descendant = ancestor.hash == last_finalized;
+				// If the best block is not a descendant of the last finalized block, it means that
+				// the best block is on a fork, that will be pruned in the future.
```

Comment:
Is this comment taking into account that the pruning logic changed? We should now prune directly when finalization takes place and not anymore when the head of a fork is below the finalized block. If we arrive in this codepath maybe the stored best block here is just below the finalized one? Logic still looks correct though.

Reply:
That makes sense, wasn't aware the pruning logic changed! I've changed the comment to reflect that, thanks!